packages/flutter_tools/lib/src/build_runner/build_runner.dart

```diff
@@ -127,57 +126,74 @@ class BuildRunner extends CodeGenerator {
     return CodeGenerationResult(packagesFile, dillFile);
   }
 
-  @override
-  Future<void> invalidateBuildScript() async {
-    final FlutterProject flutterProject = await FlutterProject.current();
-    final File buildScript = flutterProject.dartTool
-        .absolute
-        .childDirectory('flutter_tool')
-        .childFile('build.dart');
-    if (!buildScript.existsSync()) {
-      return;
-    }
-    await buildScript.delete();
-  }
-
   @override
   Future<void> generateBuildScript() async {
     final FlutterProject flutterProject = await FlutterProject.current();
-    final String generatedDirectory = fs.path.join(flutterProject.dartTool.path, 'flutter_tool');
-    final String resultScriptPath = fs.path.join(flutterProject.dartTool.path, 'build', 'entrypoint', 'build.dart');
-    if (fs.file(resultScriptPath).existsSync()) {
-      return;
+    final Directory entrypointDirectory = fs.directory(fs.path.join(flutterProject.dartTool.path, 'build', 'entrypoint'));
+    final Directory generatedDirectory = fs.directory(fs.path.join(flutterProject.dartTool.path, 'flutter_tool'));
+    final File buildScript = entrypointDirectory.childFile('build.dart');
+    final File buildSnapshot = entrypointDirectory.childFile('build.dart.snapshot');
+    final File scriptIdFile = entrypointDirectory.childFile('id');
+    final File syntheticPubspec = generatedDirectory.childFile('pubspec.yaml');
+
+    // Check if contents of builders changed. If so, invalidate build script
+    // and regnerate.
+    final YamlMap builders = await flutterProject.builders;
+    final String newContents = _produceScriptId(builders);
```

Comment:
This code deals with a large number of files. I think a more specific name for this variable could improve readability. "contents" sounds too generic. It's unclear which file it comes from.